Fixes #1560 - Migrate clay-charts to typescript

bryceosterhaus opened the pull request:
Wanted to send this just to get another pair of eyes. There are still some tests and type fixes I want to add, but I thought this was a start. I also decided to remove the react-billboardjs dependency and roll our own. By doing this it allows us to not rely on someone else's packages and also gives us more leverage with TS and using the billboard.js types.

Let me know if anything catches your eye.

matuzalemsteles commented on packages/clay-charts/src/BillboardWrapper.tsx:
```diff
@@ -0,0 +1,75 @@
+	elementProps?: HTMLAttributes<HTMLDivElement>;
+}
+
+class BillboardWrapper extends React.Component<IProps, {}> {
```
I'd say the same here, would not it be better to be a hook?

bryceosterhaus replied:
same as above

bryceosterhaus commented on packages/clay-charts/src/index.tsx:
```diff
@@ -1,31 +1,32 @@
 );
 
-export const AreaLineChart = setType('area-line');
+export const AreaLineChart = setType('area-line-range');
```
Usually I think we namespace with the `Component.***` for sub-components. Such as `Card.Header`. These charts are technically the high-level components, so it may make sense to change them to `ClayChartBar`, or we remove these specific named components and just opt that people always pass in the type to `ClayChart`.

matuzalemsteles replied:
Hm yeah, makes sense. I think it's worth delivering just the `ClayChart` and removing the other components.

bryceosterhaus replied:
Yeah I agree, I think before added this to keep the API the same as clay-charts in soy.

bryceosterhaus commented on packages/clay-charts/src/GeoMap.tsx:
```diff
@@ -82,7 +104,7 @@ class GeomapBase {
 		this._onDataLoadHandler = this._onDataLoad.bind(this);
 
 		resolveData(this._data)
```
Oh good idea. I originally just copied this straight over from the old chart component. I'll look into this improvement.

bryceosterhaus commented on packages/clay-charts/src/GeoMap.tsx:
```diff
@@ -3,34 +3,59 @@
  * Geomap Base class.
- * @class Geomap
  */
 class GeomapBase {
```
I left it as a class since it was a straight copy from the old clay-charts. I will update.

matuzalemsteles replied:
np, I figured that out. I was just wondering if you had any problems if you try

bryceosterhaus commented:
> The team of the CEREBRO suffered a lot with the charts in the past

My plan for this was to get the initial TS version merged, and then I was gonna start up Analytics and drop it in and see what might be needed. I can talk to some of the Analytics team here in LA too and figure out anything they might need. Ill do this sometime after June 17th(on PTO)

matuzalemsteles commented:
> My plan for this was to get the initial TS version merged, and then I was gonna start up Analytics and drop it in and see what might be needed. I can talk to some of the Analytics team here in LA too and figure out anything they might need. Ill do this sometime after June 17th(on PTO)

Yeah, np this is something we can go to work on and test with them, it would not be something for now.

bryceosterhaus commented:
@matuzalemsteles went ahead and updated the PR. GeoMap is actually using hooks, but still uses a vanilla js class, we can try to remove that later though.

I also didn't add the new dataProvider yet. If you want to go ahead and add that, feel free, otherwise I will add it when I am back from PTO.

matuzalemsteles commented:
hey @bryceosterhaus sorry for the delay, I ended up having medical problems last week, but I'm back today. I will look into this today with your recent changes.

> I also didn't add the new dataProvider yet. If you want to go ahead and add that, feel free, otherwise I will add it when I am back from PTO.

I was confused here, but are you thinking of adding this direct dependency to ClayCharts? I was thinking of letting the dev compose this by adding the DataProvider only when it's needed.

bryceosterhaus commented:
@matuzalemsteles I went ahead and just removed "resolveData" for now. A user should be able to just use the data provider or their own solution and then just pass in the data.

I also went ahead and added the forwardRef mentioned in issue #2064

matuzalemsteles commented:
It looks good @bryceosterhaus, I think we could try launching it along with `3.0.0-milestone-1` to test on Analytics maybe, what do you think? I think Workflow will also start working with charts. Coming soon.

bryceosterhaus commented:
@matuzalemsteles I'm gonna get analytics running locally and test out charts on it today, just in case I have missed something. If it looks okay, I'll go ahead and change this from "Draft" status and let  you know.

bryceosterhaus commented:
@matuzalemsteles I updated to point the ref to the chart instance and also updated the migration guide. Let me know if you see any issues.

bryceosterhaus commented:
@matuzalemsteles I cleaned up the tests a bit. Can you just verify the data-provider changes are fine?
